binmap/__init__.py

```diff
@@ -35,191 +44,171 @@ def __set__(self, obj, value):
-        keys = clsobject._datafields.keys()
-        sig = Signature(
-            [
-                Parameter(
-                    "binarydata",
-                    Parameter.POSITIONAL_OR_KEYWORD,
-                    default=Parameter.default,
-                )
-            ]
-            + [
-                Parameter(name, Parameter.KEYWORD_ONLY, default=Parameter.default)
-                for name in keys
-            ]
-        )
-        setattr(clsobject, "__signature__", sig)
-        for enum in clsobject._enums:
-            for value, const in clsobject._enums[enum].items():
-                if hasattr(clsobject, const.upper()):
-                    raise ValueError(f"{const} already defined")
-                setattr(clsobject, const.upper(), value)
-        for name in keys:
-            if name.startswith("_pad"):
-                setattr(clsobject, name, PaddingField(name=name))
-            elif name in clsobject._constants:
-                setattr(clsobject, name, ConstField(name=name))
-            elif name in clsobject._enums:
-                setattr(clsobject, name, EnumField(name=name))
-                setattr(clsobject, f"_{name}", BinField(name=f"_{name}"))
-            else:
-                setattr(clsobject, name, BinField(name=name))
-        return clsobject
-
-
-class Binmap(metaclass=BinmapMetaclass):
-    """A class that maps to and from binary data using :py:class:`struct`.
-    All fields in :attr:`binmap.Binmap._datafields` gets a corresponding
-    variable in instance, except variables starting with ``_pad``, which is
-    just padding.
-
-    To create an enum mapping you could add :attr:`binmap.Binmap._enums` with a
-    map to your corresponing datafield:
-
-    .. code-block:: python
-
-        class TempWind(Binmap):
-            _datafields = {"temp": "b", "wind": "B"}
-            _enums = {"wind": {0: "North", 1: "East", 2: "South", 4: "West"}}
-
-        tw = TempWind()
-        tw.temp = 3
-        tw.wind = "South"
-        print(bytes(tw))
-        b'\\x03\\x02'
-
-
-    """
-
-    #: _byteorder: charcter with byteorder
-    _byteorder = ">"
-    #: _datafields: dict with variable name as key and :py:class:`struct` `format strings`_ as value
-    _datafields = {}
-    #: _enums: dict of dicts containing maps of strings
-    _enums = {}
-    #: _constants: dict of constants. This creates a variable that is allways
-    #: the same value. It won't accept binary data with any other value
-    _constants = {}
-
-    def __init__(self, *args, **kwargs):
-        self._formatstring = self._byteorder
-        for fmt in self._datafields.values():
-            self._formatstring += fmt
-
-        bound = self.__signature__.bind(*args, **kwargs)
+        if self.name in obj.__dict__:
+            raise AttributeError(f"{self.name} is a constant")
+        else:
+            obj.__dict__.update({self.name: value})
+
+
+char = NewType("char", int)
```

Comment:
Hmm, you're right that constants usually are all caps, but since these are types I think it looks better and more consistent with other types to have them lower case. Also, already moved them to their own file in 41cffd405c85807b567146b677ea6fda65f96a65 and 1d2d1092568662254a70e4a21d16e8a4694d4fc5